File: scripts/enum/ec2.js
'use strict';

const {
  EC2Client,
  DescribeInstancesCommand,
  DescribeSecurityGroupsCommand,
  DescribeVpcsCommand,
  DescribeSnapshotsCommand,
  DescribeSnapshotAttributeCommand,
} = require('@aws-sdk/client-ec2');

const {
  ElasticLoadBalancingV2Client,
  DescribeLoadBalancersCommand: DescribeALBsCommand,
  DescribeListenersCommand,
} = require('@aws-sdk/client-elastic-load-balancing-v2');

const {
  ElasticLoadBalancingClient,
  DescribeLoadBalancersCommand: DescribeClassicLBsCommand,
} = require('@aws-sdk/client-elastic-load-balancing');

const { withRetry, paginate, createLogger } = require('../lib');
const { baseEnum } = require('../lib/base-enum');

// --- Instance enumeration ---

async function enumerateInstances(ec2, region, logger) {
  logger.log('api_call', 'DescribeInstances', { service: 'ec2' });
  const reservations = await paginate(ec2, DescribeInstancesCommand, 'Reservations');

  const findings = [];
  for (const reservation of reservations) {
    for (const instance of reservation.Instances || []) {
      const imdsVersion = instance.MetadataOptions?.HttpTokens === 'required' ? 'v2' : 'v1';
      const isImdsV1 = imdsVersion === 'v1';

      const nameTag = (instance.Tags || []).find((t) => t.Key === 'Name');
      const sgIds = (instance.SecurityGroups || []).map((sg) => sg.GroupId);

      const finding = {
        resource_type: 'ec2_instance',
        resource_id: instance.InstanceId,
        arn: `arn:aws:ec2:${region}:*:instance/${instance.InstanceId}`,
        region,
        name: nameTag ? nameTag.Value : null,
        state: instance.State?.Name || null,
        instance_type: instance.InstanceType || null,
        platform: instance.Platform || 'linux',
        public_ip: instance.PublicIpAddress || null,
        private_ip: instance.PrivateIpAddress || null,
        vpc_id: instance.VpcId || null,
        subnet_id: instance.SubnetId || null,
        iam_instance_profile: instance.IamInstanceProfile
          ? { arn: instance.IamInstanceProfile.Arn, id: instance.IamInstanceProfile.Id }
          : null,
        security_groups: sgIds,
        imds_version: imdsVersion,
        metadata_options: {
          http_tokens: instance.MetadataOptions?.HttpTokens || null,
          http_endpoint: instance.MetadataOptions?.HttpEndpoint || null,
          http_put_response_hop_limit: instance.MetadataOptions?.HttpPutResponseHopLimit || null,
        },
        findings: [],
      };

      if (isImdsV1) {
        finding.findings.push({
          type: 'imds_v1_enabled',
          severity: 'critical',
          detail: 'Instance uses IMDSv1 (HttpTokens=optional) — credential theft via SSRF',
        });
      }

      if (instance.PublicIpAddress) {
        finding.findings.push({
          type: 'public_ip',
          severity: 'info',
          detail: `Instance has public IP: ${instance.PublicIpAddress}`,
        });
      }

      if (!instance.IamInstanceProfile) {
        finding.findings.push({
          type: 'no_instance_profile',
          severity: 'info',
          detail: 'No IAM instance profile attached',
        });
      }

      findings.push(finding);
    }
  }

  logger.log('info', 'DescribeInstances_Complete', { count: findings.length });
  return findings;
}

// --- Security Groups ---

async function enumerateSecurityGroups(ec2, region, logger) {
  logger.log('api_call', 'DescribeSecurityGroups', { service: 'ec2' });
  const groups = await paginate(ec2, DescribeSecurityGroupsCommand, 'SecurityGroups');

  const findings = [];
  for (const sg of groups) {
    const inboundRules = (sg.IpPermissions || []).map((perm) => {
      const sources = [
        ...(perm.IpRanges || []).map((r) => r.CidrIp),
        ...(perm.Ipv6Ranges || []).map((r) => r.CidrIpv6),
        ...(perm.PrefixListIds || []).map((p) => p.PrefixListId),
        ...(perm.UserIdGroupPairs || []).map((g) => g.GroupId),
      ];
      return {
        protocol: perm.IpProtocol || null,
        from_port: perm.FromPort ?? null,
        to_port: perm.ToPort ?? null,
        sources,
      };
    });

    const openToWorld = inboundRules.some((rule) =>
      rule.sources.some((s) => s === '0.0.0.0/0' || s === '::/0')
    );

    const finding = {
      resource_type: 'ec2_security_group',
      resource_id: sg.GroupId,
      arn: `arn:aws:ec2:${region}:*:security-group/${sg.GroupId}`,
      region,
      name: sg.GroupName || null,
      description: sg.Description || null,
      vpc_id: sg.VpcId || null,
      inbound_rules: inboundRules,
      findings: [],
    };

    if (openToWorld) {
      finding.findings.push({
        type: 'open_to_world',
        severity: 'high',
        detail: 'Security group has inbound rule(s) open to 0.0.0.0/0 or ::/0',
      });
    }

    findings.push(finding);
  }

  logger.log('info', 'DescribeSecurityGroups_Complete', { count: findings.length });
  return findings;
}

// --- VPCs ---

async function enumerateVpcs(ec2, region, logger) {
  logger.log('api_call', 'DescribeVpcs', { service: 'ec2' });
  const vpcs = await paginate(ec2, DescribeVpcsCommand, 'Vpcs');

  const findings = [];
  for (const vpc of vpcs) {
    const nameTag = (vpc.Tags || []).find((t) => t.Key === 'Name');
    findings.push({
      resource_type: 'ec2_vpc',
      resource_id: vpc.VpcId,
      arn: `arn:aws:ec2:${region}:*:vpc/${vpc.VpcId}`,
      region,
      name: nameTag ? nameTag.Value : null,
      cidr_block: vpc.CidrBlock || null,
      is_default: vpc.IsDefault || false,
      state: vpc.State || null,
      findings: [],
    });
  }

  logger.log('info', 'DescribeVpcs_Complete', { count: findings.length });
  return findings;
}

// --- Snapshots ---

async function enumerateSnapshots(ec2, accountId, region, logger) {
  logger.log('api_call', 'DescribeSnapshots', { service: 'ec2', owner: 'self' });
  const snapshots = await paginate(ec2, DescribeSnapshotsCommand, 'Snapshots', {
    params: { OwnerIds: ['self'] },
  });

  const findings = [];
  for (const snap of snapshots) {
    let isPublic = false;

    try {
      logger.log('api_call', 'DescribeSnapshotAttribute', { snapshot: snap.SnapshotId });
      const attrResp = await withRetry(() =>
        ec2.send(new DescribeSnapshotAttributeCommand({
          SnapshotId: snap.SnapshotId,
          Attribute: 'createVolumePermission',
        }))
      );
      const perms = attrResp.CreateVolumePermissions || [];
      isPublic = perms.some((p) => p.Group === 'all');
    } catch (err) {
      logger.log('warning', 'DescribeSnapshotAttribute', {
        snapshot: snap.SnapshotId,
        error: err.message,
      });
    }

    const finding = {
      resource_type: 'ec2_snapshot',
      resource_id: snap.SnapshotId,
      arn: `arn:aws:ec2:${region}:${accountId}:snapshot/${snap.SnapshotId}`,
      region,
      volume_id: snap.VolumeId || null,
      volume_size_gb: snap.VolumeSize || null,
      encrypted: snap.Encrypted || false,
      state: snap.State || null,
      is_public: isPublic,
      findings: [],
    };

    if (isPublic) {
      finding.findings.push({
        type: 'public_snapshot',
        severity: 'critical',
        detail: 'Snapshot is publicly shared (createVolumePermission includes "all")',
      });
    }

    if (!snap.Encrypted) {
      finding.findings.push({
        type: 'unencrypted_snapshot',
        severity: 'medium',
        detail: 'Snapshot is not encrypted',
      });
    }

    findings.push(finding);
  }

  logger.log('info', 'DescribeSnapshots_Complete', { count: findings.length });
  return findings;
}

// --- ELBv2 (ALB/NLB) ---

async function enumerateELBv2(elbv2, region, logger) {
  logger.log('api_call', 'DescribeLoadBalancers_v2', { service: 'elbv2' });
  const lbs = await paginate(elbv2, DescribeALBsCommand, 'LoadBalancers', {
    tokenKey: 'Marker',
    responseTokenKey: 'NextMarker',
  });

  const findings = [];
  for (const lb of lbs) {
    let listeners = [];
    try {
      logger.log('api_call', 'DescribeListeners', { lb_arn: lb.LoadBalancerArn });
      listeners = await paginate(elbv2, DescribeListenersCommand, 'Listeners', {
        params: { LoadBalancerArn: lb.LoadBalancerArn },
        tokenKey: 'Marker',
        responseTokenKey: 'NextMarker',
      });
    } catch (err) {
      logger.log('warning', 'DescribeListeners', { lb_arn: lb.LoadBalancerArn, error: err.message });
    }

    const finding = {
      resource_type: 'ec2_load_balancer',
      resource_id: lb.LoadBalancerName,
      arn: lb.LoadBalancerArn,
      region,
      type: lb.Type || null,
      scheme: lb.Scheme || null,
      state: lb.State?.Code || null,
      dns_name: lb.DNSName || null,
      vpc_id: lb.VpcId || null,
      availability_zones: (lb.AvailabilityZones || []).map((az) => az.ZoneName),
      listeners: listeners.map((l) => ({
        port: l.Port,
        protocol: l.Protocol,
        ssl_policy: l.SslPolicy || null,
      })),
      findings: [],
    };

    if (lb.Scheme === 'internet-facing') {
      finding.findings.push({
        type: 'internet_facing',
        severity: 'info',
        detail: `Load balancer is internet-facing: ${lb.DNSName}`,
      });
    }

    findings.push(finding);
  }

  logger.log('info', 'DescribeLoadBalancers_v2_Complete', { count: findings.length });
  return findings;
}

// --- Classic ELB ---

async function enumerateClassicELB(elb, region, logger) {
  logger.log('api_call', 'DescribeLoadBalancers_classic', { service: 'elb' });
  const lbs = await paginate(elb, DescribeClassicLBsCommand, 'LoadBalancerDescriptions', {
    tokenKey: 'Marker',
    responseTokenKey: 'NextMarker',
  });

  const findings = [];
  for (const lb of lbs) {
    const finding = {
      resource_type: 'ec2_load_balancer',
      resource_id: lb.LoadBalancerName,
      arn: null, // Classic ELBs don't have ARN in describe response
      region,
      type: 'classic',
      scheme: lb.Scheme || null,
      dns_name: lb.DNSName || null,
      vpc_id: lb.VPCId || null,
      availability_zones: lb.AvailabilityZones || [],
      listeners: (lb.ListenerDescriptions || []).map((ld) => ({
        port: ld.Listener?.LoadBalancerPort || null,
        protocol: ld.Listener?.Protocol || null,
        instance_port: ld.Listener?.InstancePort || null,
        instance_protocol: ld.Listener?.InstanceProtocol || null,
      })),
      findings: [],
    };

    if (lb.Scheme === 'internet-facing') {
      finding.findings.push({
        type: 'internet_facing',
        severity: 'info',
        detail: `Classic load balancer is internet-facing: ${lb.DNSName}`,
      });
    }

    findings.push(finding);
  }

  logger.log('info', 'DescribeLoadBalancers_classic_Complete', { count: findings.length });
  return findings;
}

// --- Run (dependency-injectable) ---

async function run(opts = {}) {
  const { runDir, region } = opts;
  const accountId = opts.accountId;

  if (!runDir || !region) {
    throw new Error('runDir and region are required');
  }

  const ec2 = opts.clients?.ec2 ?? new EC2Client({ region });
  const elbv2 = opts.clients?.elbv2 ?? new ElasticLoadBalancingV2Client({ region });
  const elb = opts.clients?.elb ?? new ElasticLoadBalancingClient({ region });

  const logger = opts.logger || createLogger(runDir, 'ec2');
  logger.log('info', 'EC2_Enumeration_Start', { region });

  const allFindings = [];
  let status = 'complete';
  const errors = [];

  // 1. Instances
  try {
    const instances = await enumerateInstances(ec2, region, logger);
    allFindings.push(...instances);
  } catch (err) {
    errors.push({ resource_type: 'ec2_instance', error: err.message });
    logger.log('error', 'DescribeInstances', { error: err.message });
  }

  // 2. Security Groups
  try {
    const sgs = await enumerateSecurityGroups(ec2, region, logger);
    allFindings.push(...sgs);
  } catch (err) {
    errors.push({ resource_type: 'ec2_security_group', error: err.message });
    logger.log('error', 'DescribeSecurityGroups', { error: err.message });
  }

  // 3. VPCs
  try {
    const vpcs = await enumerateVpcs(ec2, region, logger);
    allFindings.push(...vpcs);
  } catch (err) {
    errors.push({ resource_type: 'ec2_vpc', error: err.message });
    logger.log('error', 'DescribeVpcs', { error: err.message });
  }

  // 4. Snapshots
  try {
    const snaps = await enumerateSnapshots(ec2, accountId, region, logger);
    allFindings.push(...snaps);
  } catch (err) {
    errors.push({ resource_type: 'ec2_snapshot', error: err.message });
    logger.log('error', 'DescribeSnapshots', { error: err.message });
  }

  // 5. ELBv2 (ALB/NLB)
  try {
    const elbv2Findings = await enumerateELBv2(elbv2, region, logger);
    allFindings.push(...elbv2Findings);
  } catch (err) {
    errors.push({ resource_type: 'ec2_load_balancer_v2', error: err.message });
    logger.log('error', 'DescribeLoadBalancers_v2', { error: err.message });
  }

  // 6. Classic ELB
  try {
    const classicFindings = await enumerateClassicELB(elb, region, logger);
    allFindings.push(...classicFindings);
  } catch (err) {
    errors.push({ resource_type: 'ec2_load_balancer_classic', error: err.message });
    logger.log('error', 'DescribeLoadBalancers_classic', { error: err.message });
  }

  if (errors.length > 0) status = 'partial';

  await logger.flush();
  return { findings: allFindings, status };
}

if (require.main === module) {
  baseEnum({ module: 'ec2', run });
}

module.exports = { run };

File: enumerators/ec2.py
from __future__ import annotations

from typing import Any

from scope_core.coverage import CoverageTracker
from scope_core.models import ModuleEnvelope


PRIMARY_CHECKS: list[str] = []
REQUIRED_CHECKS = [
    "describe_instances",
    "describe_security_groups",
    "describe_vpcs",
    "describe_snapshots",
    "snapshot_attributes",
    "describe_load_balancers_v2",
    "describe_listeners",
    "describe_load_balancers_classic",
]


def run(factory: Any, region: str) -> ModuleEnvelope:
    tracker = CoverageTracker(primary=PRIMARY_CHECKS, required=REQUIRED_CHECKS)
    ec2 = factory.client("ec2")
    elbv2 = factory.client("elbv2")
    elb = factory.client("elb")
    findings: list[dict[str, Any]] = []

    findings.extend(_enumerate_instances(factory, ec2, region, tracker))
    findings.extend(_enumerate_security_groups(factory, ec2, region, tracker))
    findings.extend(_enumerate_vpcs(factory, ec2, region, tracker))
    findings.extend(_enumerate_snapshots(factory, ec2, factory.account_id, region, tracker))
    findings.extend(_enumerate_elbv2(factory, elbv2, region, tracker))
    findings.extend(_enumerate_classic_elb(factory, elb, region, tracker))

    coverage, errors = tracker.to_envelope_fields()
    return ModuleEnvelope(
        module="ec2",
        account_id=factory.account_id,
        region=region,
        status=tracker.derive_status(),
        resources=findings,
        coverage=coverage,
        errors=errors,
    )


def _enumerate_instances(factory: Any, ec2: Any, region: str, tracker: CoverageTracker) -> list[dict[str, Any]]:
    try:
        reservations = factory.paginate(ec2, "describe_instances", "Reservations")
        tracker.record_ok("describe_instances", resource="instances")
    except Exception as err:
        tracker.record_resource_failure("describe_instances", None, "ec2.DescribeInstances", err)
        return []

    findings: list[dict[str, Any]] = []
    for reservation in reservations:
        for instance in reservation.get("Instances") or []:
            metadata = instance.get("MetadataOptions") or {}
            imds_version = "v2" if metadata.get("HttpTokens") == "required" else "v1"
            name_tag = next((tag for tag in instance.get("Tags") or [] if tag.get("Key") == "Name"), None)
            finding = {
                "resource_type": "ec2_instance",
                "resource_id": instance.get("InstanceId"),
                "arn": f"arn:aws:ec2:{region}:*:instance/{instance.get('InstanceId')}",
                "region": region,
                "name": name_tag.get("Value") if name_tag else None,
                "state": (instance.get("State") or {}).get("Name"),
                "instance_type": instance.get("InstanceType"),
                "platform": instance.get("Platform") or "linux",
                "public_ip": instance.get("PublicIpAddress"),
                "private_ip": instance.get("PrivateIpAddress"),
                "vpc_id": instance.get("VpcId"),
                "subnet_id": instance.get("SubnetId"),
                "iam_instance_profile": _instance_profile(instance.get("IamInstanceProfile")),
                "security_groups": [sg.get("GroupId") for sg in instance.get("SecurityGroups") or []],
                "imds_version": imds_version,
                "metadata_options": {
                    "http_tokens": metadata.get("HttpTokens"),
                    "http_endpoint": metadata.get("HttpEndpoint"),
                    "http_put_response_hop_limit": metadata.get("HttpPutResponseHopLimit"),
                },
                "findings": [],
            }
            if imds_version == "v1":
                finding["findings"].append(
                    {
                        "type": "imds_v1_enabled",
                        "severity": "critical",
                        "detail": "Instance uses IMDSv1 (HttpTokens=optional) — credential theft via SSRF",
                    }
                )
            if instance.get("PublicIpAddress"):
                finding["findings"].append(
                    {
                        "type": "public_ip",
                        "severity": "info",
                        "detail": f"Instance has public IP: {instance.get('PublicIpAddress')}",
                    }
                )
            if not instance.get("IamInstanceProfile"):
                finding["findings"].append(
                    {
                        "type": "no_instance_profile",
                        "severity": "info",
                        "detail": "No IAM instance profile attached",
                    }
                )
            findings.append(finding)
    return findings


def _instance_profile(profile: dict[str, Any] | None) -> dict[str, Any] | None:
    if not profile:
        return None
    return {"arn": profile.get("Arn"), "id": profile.get("Id")}


def _enumerate_security_groups(factory: Any, ec2: Any, region: str, tracker: CoverageTracker) -> list[dict[str, Any]]:
    try:
        groups = factory.paginate(ec2, "describe_security_groups", "SecurityGroups")
        tracker.record_ok("describe_security_groups", resource="security_groups")
    except Exception as err:
        tracker.record_resource_failure("describe_security_groups", None, "ec2.DescribeSecurityGroups", err)
        return []

    findings: list[dict[str, Any]] = []
    for group in groups:
        inbound_rules = [_inbound_rule(permission) for permission in group.get("IpPermissions") or []]
        finding = {
            "resource_type": "ec2_security_group",
            "resource_id": group.get("GroupId"),
            "arn": f"arn:aws:ec2:{region}:*:security-group/{group.get('GroupId')}",
            "region": region,
            "name": group.get("GroupName"),
            "description": group.get("Description"),
            "vpc_id": group.get("VpcId"),
            "inbound_rules": inbound_rules,
            "findings": [],
        }
        if any(source in {"0.0.0.0/0", "::/0"} for rule in inbound_rules for source in rule["sources"]):
            finding["findings"].append(
                {
                    "type": "open_to_world",
                    "severity": "high",
                    "detail": "Security group has inbound rule(s) open to 0.0.0.0/0 or ::/0",
                }
            )
        findings.append(finding)
    return findings


def _inbound_rule(permission: dict[str, Any]) -> dict[str, Any]:
    sources = [
        *[item.get("CidrIp") for item in permission.get("IpRanges") or []],
        *[item.get("CidrIpv6") for item in permission.get("Ipv6Ranges") or []],
        *[item.get("PrefixListId") for item in permission.get("PrefixListIds") or []],
        *[item.get("GroupId") for item in permission.get("UserIdGroupPairs") or []],
    ]
    return {
        "protocol": permission.get("IpProtocol"),
        "from_port": permission.get("FromPort"),
        "to_port": permission.get("ToPort"),
        "sources": [source for source in sources if source is not None],
    }


def _enumerate_vpcs(factory: Any, ec2: Any, region: str, tracker: CoverageTracker) -> list[dict[str, Any]]:
    try:
        vpcs = factory.paginate(ec2, "describe_vpcs", "Vpcs")
        tracker.record_ok("describe_vpcs", resource="vpcs")
    except Exception as err:
        tracker.record_resource_failure("describe_vpcs", None, "ec2.DescribeVpcs", err)
        return []

    findings = []
    for vpc in vpcs:
        name_tag = next((tag for tag in vpc.get("Tags") or [] if tag.get("Key") == "Name"), None)
        findings.append(
            {
                "resource_type": "ec2_vpc",
                "resource_id": vpc.get("VpcId"),
                "arn": f"arn:aws:ec2:{region}:*:vpc/{vpc.get('VpcId')}",
                "region": region,
                "name": name_tag.get("Value") if name_tag else None,
                "cidr_block": vpc.get("CidrBlock"),
                "is_default": vpc.get("IsDefault") or False,
                "state": vpc.get("State"),
                "findings": [],
            }
        )
    return findings


def _enumerate_snapshots(factory: Any, ec2: Any, account_id: str, region: str, tracker: CoverageTracker) -> list[dict[str, Any]]:
    try:
        snapshots = factory.paginate(ec2, "describe_snapshots", "Snapshots", OwnerIds=["self"])
        tracker.record_ok("describe_snapshots", resource="snapshots")
    except Exception as err:
        tracker.record_resource_failure("describe_snapshots", None, "ec2.DescribeSnapshots", err)
        return []

    findings: list[dict[str, Any]] = []
    for snapshot in snapshots:
        snapshot_arn = f"arn:aws:ec2:{region}:{account_id}:snapshot/{snapshot.get('SnapshotId')}"
        is_public = False
        snapshot_attributes_status = None
        try:
            response = ec2.describe_snapshot_attribute(
                SnapshotId=snapshot.get("SnapshotId"),
                Attribute="createVolumePermission",
            )
            permissions = response.get("CreateVolumePermissions") or []
            is_public = any(permission.get("Group") == "all" for permission in permissions)
            snapshot_attributes_status = "present"
            tracker.record_ok("snapshot_attributes", resource=snapshot_arn)
        except Exception as err:
            code = _error_code(err)
            snapshot_attributes_status = _classify_status(code)
            tracker.record_resource_failure(
                "snapshot_attributes",
                snapshot_arn,
                "ec2.DescribeSnapshotAttribute",
                err,
            )

        finding = {
            "resource_type": "ec2_snapshot",
            "resource_id": snapshot.get("SnapshotId"),
            "arn": snapshot_arn,
            "region": region,
            "volume_id": snapshot.get("VolumeId"),
            "volume_size_gb": snapshot.get("VolumeSize"),
            "encrypted": snapshot.get("Encrypted") or False,
            "state": snapshot.get("State"),
            "is_public": is_public,
            "snapshot_attributes_status": snapshot_attributes_status,
            "findings": [],
        }
        if is_public:
            finding["findings"].append(
                {
                    "type": "public_snapshot",
                    "severity": "critical",
                    "detail": 'Snapshot is publicly shared (createVolumePermission includes "all")',
                }
            )
        if not snapshot.get("Encrypted"):
            finding["findings"].append(
                {
                    "type": "unencrypted_snapshot",
                    "severity": "medium",
                    "detail": "Snapshot is not encrypted",
                }
            )
        findings.append(finding)
    return findings


def _enumerate_elbv2(factory: Any, elbv2: Any, region: str, tracker: CoverageTracker) -> list[dict[str, Any]]:
    try:
        load_balancers = factory.paginate(elbv2, "describe_load_balancers", "LoadBalancers")
        tracker.record_ok("describe_load_balancers_v2", resource="load_balancers_v2")
    except Exception as err:
        tracker.record_resource_failure("describe_load_balancers_v2", None, "elbv2.DescribeLoadBalancers", err)
        return []

    findings: list[dict[str, Any]] = []
    for load_balancer in load_balancers:
        lb_arn = load_balancer.get("LoadBalancerArn")
        listeners: list[dict[str, Any]] = []
        listeners_status = None
        try:
            raw_listeners = factory.paginate(elbv2, "describe_listeners", "Listeners", LoadBalancerArn=lb_arn)
            listeners = [
                {
                    "port": listener.get("Port"),
                    "protocol": listener.get("Protocol"),
                    "ssl_policy": listener.get("SslPolicy"),
                }
                for listener in raw_listeners
            ]
            listeners_status = "present"
            tracker.record_ok("describe_listeners", resource=lb_arn)
        except Exception as err:
            listeners_status = _classify_status(_error_code(err))
            tracker.record_resource_failure("describe_listeners", lb_arn, "elbv2.DescribeListeners", err)

        finding = {
            "resource_type": "ec2_load_balancer",
            "resource_id": load_balancer.get("LoadBalancerName"),
            "arn": lb_arn,
            "region": region,
            "type": load_balancer.get("Type"),
            "scheme": load_balancer.get("Scheme"),
            "state": (load_balancer.get("State") or {}).get("Code"),
            "dns_name": load_balancer.get("DNSName"),
            "vpc_id": load_balancer.get("VpcId"),
            "availability_zones": [az.get("ZoneName") for az in load_balancer.get("AvailabilityZones") or []],
            "listeners": listeners,
            "listeners_status": listeners_status,
            "findings": [],
        }
        if load_balancer.get("Scheme") == "internet-facing":
            finding["findings"].append(
                {
                    "type": "internet_facing",
                    "severity": "info",
                    "detail": f"Load balancer is internet-facing: {load_balancer.get('DNSName')}",
                }
            )
        findings.append(finding)
    return findings


def _enumerate_classic_elb(factory: Any, elb: Any, region: str, tracker: CoverageTracker) -> list[dict[str, Any]]:
    try:
        load_balancers = factory.paginate(elb, "describe_load_balancers", "LoadBalancerDescriptions")
        tracker.record_ok("describe_load_balancers_classic", resource="load_balancers_classic")
    except Exception as err:
        tracker.record_resource_failure("describe_load_balancers_classic", None, "elb.DescribeLoadBalancers", err)
        return []

    findings = []
    for load_balancer in load_balancers:
        finding = {
            "resource_type": "ec2_load_balancer",
            "resource_id": load_balancer.get("LoadBalancerName"),
            "arn": None,
            "region": region,
            "type": "classic",
            "scheme": load_balancer.get("Scheme"),
            "dns_name": load_balancer.get("DNSName"),
            "vpc_id": load_balancer.get("VPCId"),
            "availability_zones": load_balancer.get("AvailabilityZones") or [],
            "listeners": [
                {
                    "port": (description.get("Listener") or {}).get("LoadBalancerPort"),
                    "protocol": (description.get("Listener") or {}).get("Protocol"),
                    "instance_port": (description.get("Listener") or {}).get("InstancePort"),
                    "instance_protocol": (description.get("Listener") or {}).get("InstanceProtocol"),
                }
                for description in load_balancer.get("ListenerDescriptions") or []
            ],
            "findings": [],
        }
        if load_balancer.get("Scheme") == "internet-facing":
            finding["findings"].append(
                {
                    "type": "internet_facing",
                    "severity": "info",
                    "detail": f"Classic load balancer is internet-facing: {load_balancer.get('DNSName')}",
                }
            )
        findings.append(finding)
    return findings


def _classify_status(code: str) -> str:
    return "access_denied" if code in {"AccessDenied", "AccessDeniedException"} else "error"


def _error_code(err: Exception) -> str:
    response = getattr(err, "response", None)
    if isinstance(response, dict):
        error = response.get("Error")
        if isinstance(error, dict) and error.get("Code"):
            return str(error["Code"])
    code = getattr(err, "Code", None)
    if code:
        return str(code)
    return err.__class__.__name__ or "Error"
